[JSC] Implement wasm exception in new BBQ

Constellation opened the pull request:
#### e2672d2623ed17421641e2827515bd6b02c893ea
<pre>
[JSC] Implement wasm exception in new BBQ
<a href="https://bugs.webkit.org/show_bug.cgi?id=252959">https://bugs.webkit.org/show_bug.cgi?id=252959</a>
rdar://105692614

Reviewed by Tadeu Zagallo.

This patch implements wasm exception in new BBQ. The implementation is just following
to how exception is implemented in the other tiers (LLInt, old BBQ Air), so not introducing a new concept.

1. Catch / CatchAll blocks should have implicit stack slot to hold exception (as the same to old Air and LLInt).
   This is represented as implicitSlots.
2. When catch / catch-all is running, we already flush everything in endBlock. Thus, we do not need to have stackmap
   at all: every wasm temps are on stack. So only thing we need is just spreading wasm exception parameters to the
   block&apos;s parameters (this is the same to what wasm LLInt exception handling is doing).
3. Each try-block and exception-throwing operations incremenets m_callSiteIndex. And emit it to
   CCallHelpers::tagFor(CallFrameSlot::argumentCountIncludingThis) slot. This number is used as a range of covering of
   the exception handler. (This is exactly the same design to DFG / FTL exception handling). And this is the same to
   old BBQ Air exception handling.
4. Entrance of exception handler is represented as a label inside code. And these labels are extracted for exception handlers.

* Source/JavaScriptCore/wasm/WasmB3IRGenerator.h:
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJIT::ControlData::ControlData):
(JSC::Wasm::BBQJIT::ControlData::endBlock):
(JSC::Wasm::BBQJIT::ControlData::startBlock):
(JSC::Wasm::BBQJIT::ControlData::resumeBlock):
(JSC::Wasm::BBQJIT::ControlData::releaseJumps):
(JSC::Wasm::BBQJIT::ControlData::implicitSlots const):
(JSC::Wasm::BBQJIT::ControlData::argumentLocations const):
(JSC::Wasm::BBQJIT::ControlData::resultLocations const):
(JSC::Wasm::BBQJIT::ControlData::setCatchKind):
(JSC::Wasm::BBQJIT::ControlData::tryStart const):
(JSC::Wasm::BBQJIT::ControlData::tryEnd const):
(JSC::Wasm::BBQJIT::ControlData::tryCatchDepth const):
(JSC::Wasm::BBQJIT::ControlData::setTryInfo):
(JSC::Wasm::BBQJIT::BBQJIT):
(JSC::Wasm::BBQJIT::topValue):
(JSC::Wasm::BBQJIT::exception):
(JSC::Wasm::BBQJIT::emitEntryTierUpCheck):
(JSC::Wasm::BBQJIT::addTopLevel):
(JSC::Wasm::BBQJIT::addBlock):
(JSC::Wasm::BBQJIT::addLoop):
(JSC::Wasm::BBQJIT::addIf):
(JSC::Wasm::BBQJIT::addElse):
(JSC::Wasm::BBQJIT::addElseToUnreachable):
(JSC::Wasm::BBQJIT::addTry):
(JSC::Wasm::BBQJIT::emitCatchPrologue):
(JSC::Wasm::BBQJIT::emitCatchAllImpl):
(JSC::Wasm::BBQJIT::emitCatchImpl):
(JSC::Wasm::BBQJIT::addCatch):
(JSC::Wasm::BBQJIT::addCatchToUnreachable):
(JSC::Wasm::BBQJIT::addCatchAll):
(JSC::Wasm::BBQJIT::addCatchAllToUnreachable):
(JSC::Wasm::BBQJIT::addDelegate):
(JSC::Wasm::BBQJIT::addDelegateToUnreachable):
(JSC::Wasm::BBQJIT::addThrow):
(JSC::Wasm::BBQJIT::addRethrow):
(JSC::Wasm::BBQJIT::prepareForExceptions):
(JSC::Wasm::BBQJIT::addReturn):
(JSC::Wasm::BBQJIT::endBlock):
(JSC::Wasm::BBQJIT::addEndToUnreachable):
(JSC::Wasm::BBQJIT::endTopLevel):
(JSC::Wasm::BBQJIT::flushRegistersForException):
(JSC::Wasm::BBQJIT::returnValuesFromCall):
(JSC::Wasm::BBQJIT::emitCCall):
(JSC::Wasm::BBQJIT::addCall):
(JSC::Wasm::BBQJIT::emitIndirectCall):
(JSC::Wasm::BBQJIT::notifyFunctionUsesSIMD):
(JSC::Wasm::BBQJIT::takeExceptionHandlers):
(JSC::Wasm::BBQJIT::takeCatchEntrypoints):
(JSC::Wasm::BBQJIT::emitShuffle):
(JSC::Wasm::parseAndCompileBBQ):
(JSC::Wasm::BBQJIT::ControlData::allocateArgumentOrResult): Deleted.
* Source/JavaScriptCore/wasm/WasmIRGeneratorHelpers.h:
(JSC::Wasm::computeExceptionHandlerAndLoopEntrypointLocations):

Canonical link: <a href="https://commits.webkit.org/260905@main">https://commits.webkit.org/260905@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b9dbeb89fad047c8323d95f76ae7fd128844f55c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/109891 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/77/builds/19000 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/42601 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/87/builds/1347 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/118972 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/76/builds/20466 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/85/builds/10194 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/102160 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/115639 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/76/builds/20466 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/42601 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/102160 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/76/builds/20466 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/42601 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/102160 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/1/builds/98721 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/11709 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/42601 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/38/builds/99632 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/9744 "Built successfully and passed tests") | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/82/builds/12347 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/85/builds/10194 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/46/builds/31152 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/17716 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/42601 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/107650 "Built successfully") | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/79/builds/14124 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | [  ~~🧪 jsc-mips-tests~~](https://ews-build.webkit.org/#/builders/45/builds/26563 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/4103 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

tadeuzagallo commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -5927,39 +5985,312 @@ class BBQJIT {
+        static_assert(noOverlap(GPRInfo::nonPreservedNonArgumentGPR0, GPRInfo::returnValueGPR, GPRInfo::returnValueGPR2));
+        bind(this->exception(dataCatch), Location::fromGPR(GPRInfo::returnValueGPR));
+        Stack emptyStack { };
+        dataCatch.startBlock(*this, emptyStack);
```
Why not keep all this in the prologue?

Constellation replied:
That sounds good. I would like to keep

```
        bind(this->exception(dataCatch), Location::fromGPR(GPRInfo::returnValueGPR));
        Stack emptyStack { };
        dataCatch.startBlock(*this, emptyStack);
```

separately since they are more tied to BBQJIT's mechanism (not just a JIT code generator).

tadeuzagallo commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -5927,39 +5985,312 @@ class BBQJIT {
+                break;
+            case TypeKind::I31ref:
+            case TypeKind::I64:
+            case TypeKind::Ref:
+            case TypeKind::RefNull:
+            case TypeKind::Arrayref:
+            case TypeKind::Structref:
+            case TypeKind::Funcref:
+            case TypeKind::Externref:
+            case TypeKind::Rec:
+            case TypeKind::Sub:
+            case TypeKind::Array:
+            case TypeKind::Struct:
+            case TypeKind::Func: {
```
Shouldn't some of these be `loadPtr`?

Constellation replied:
Since they are EncodedJSValue, it should be load64.

tadeuzagallo commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -5927,39 +5985,312 @@ class BBQJIT {
+    PartialResult WARN_UNUSED_RETURN addCatchAll(Stack& expressionStack, ControlType& data)
+    {
+        m_usesExceptions = true;
+        data.endBlock(*this, data, expressionStack, false, true);
```
Just a nit, and I could be missing something, but seems like you could `endBlock` here and then share the rest of the implementation with `addCatchAllToUnreachable` by passing an optional jump? (same goes to catch)

Constellation replied:
For now, I would like to keep it streamlined simple one. After OSR entry thing is done and once we see the final implementation, I would like to do some refactoring (until that, simple duplicate code is better for modifying some part of it, since anyway OSR entry code also need to touch some of exception code.).

tadeuzagallo commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -5927,39 +5985,312 @@ class BBQJIT {
+            Location stackLocation = Location::fromStackArgument(i * sizeof(uint64_t));
+            Value argument = arguments[i];
+            if (argument.type() == TypeKind::V128)
+                emitMove(Value::fromF64(0), stackLocation);
```
I don't understand, why 0 here?

Constellation replied:
SIMD cannot be handled by wasm exception right now (in spec and implementation), and the value gets automatically zero.

tadeuzagallo commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -6290,7 +6645,7 @@ class BBQJIT {
 
         // FIXME: Probably we should make CCall more lower level, and we should bind the result to Value separately.
-        result = Value::fromTemp(returnType, currentControlData().enclosedHeight() + m_parser->expressionStack().size());
+        result = Value::fromTemp(returnType, currentControlData().enclosedHeight() + currentControlData().implicitSlots() + m_parser->expressionStack().size());
```
nit: All these modifications just adding `implicitSlots` make me think it'd be nice to have a helper function

Constellation replied:
Right now, I would like to keep it since some place explicitly do not require implicitSlots. So every places, we should consider carefully about whether it is right. And helper function will hide this.

tadeuzagallo reviewed (approved):
r=me. I left a few comments, but it's mostly nits and questions, feel free to ignore.

ddegazio commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -5927,39 +5985,312 @@ class BBQJIT {
+
+        JIT_COMMENT(m_jit, "Configure wasm context instance");
+        m_jit.loadPtr(CCallHelpers::Address(GPRInfo::callFrameRegister, CallFrameSlot::codeBlock * sizeof(Register)), GPRInfo::wasmContextInstancePointer);
+        m_frameSizeLabels.append(m_jit.moveWithPatch(TrustedImmPtr(nullptr), GPRInfo::regT0));
```
Is there a reason we need to patch catch-block prologues? Maybe we can emit these at the end once we know `m_frameSize`?

Constellation replied:
No reason right now. Just a quick implementation to make it work first. My plan is doing refactoring etc. after all the features are implemented and once we enabled & start testing. Until that, I tend to make it as simple as possible.

Constellation committed:
[JSC] Implement wasm exception in new BBQ
https://bugs.webkit.org/show_bug.cgi?id=252959
rdar://105692614

Reviewed by Tadeu Zagallo.

This patch implements wasm exception in new BBQ. The implementation is just following
to how exception is implemented in the other tiers (LLInt, old BBQ Air), so not introducing a new concept.

1. Catch / CatchAll blocks should have implicit stack slot to hold exception (as the same to old Air and LLInt).
   This is represented as implicitSlots.
2. When catch / catch-all is running, we already flush everything in endBlock. Thus, we do not need to have stackmap
   at all: every wasm temps are on stack. So only thing we need is just spreading wasm exception parameters to the
   block's parameters (this is the same to what wasm LLInt exception handling is doing).
3. Each try-block and exception-throwing operations incremenets m_callSiteIndex. And emit it to
   CCallHelpers::tagFor(CallFrameSlot::argumentCountIncludingThis) slot. This number is used as a range of covering of
   the exception handler. (This is exactly the same design to DFG / FTL exception handling). And this is the same to
   old BBQ Air exception handling.
4. Entrance of exception handler is represented as a label inside code. And these labels are extracted for exception handlers.

* Source/JavaScriptCore/wasm/WasmB3IRGenerator.h:
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJIT::ControlData::ControlData):
(JSC::Wasm::BBQJIT::ControlData::endBlock):
(JSC::Wasm::BBQJIT::ControlData::startBlock):
(JSC::Wasm::BBQJIT::ControlData::resumeBlock):
(JSC::Wasm::BBQJIT::ControlData::releaseJumps):
(JSC::Wasm::BBQJIT::ControlData::implicitSlots const):
(JSC::Wasm::BBQJIT::ControlData::argumentLocations const):
(JSC::Wasm::BBQJIT::ControlData::resultLocations const):
(JSC::Wasm::BBQJIT::ControlData::setCatchKind):
(JSC::Wasm::BBQJIT::ControlData::tryStart const):
(JSC::Wasm::BBQJIT::ControlData::tryEnd const):
(JSC::Wasm::BBQJIT::ControlData::tryCatchDepth const):
(JSC::Wasm::BBQJIT::ControlData::setTryInfo):
(JSC::Wasm::BBQJIT::BBQJIT):
(JSC::Wasm::BBQJIT::topValue):
(JSC::Wasm::BBQJIT::exception):
(JSC::Wasm::BBQJIT::emitEntryTierUpCheck):
(JSC::Wasm::BBQJIT::addTopLevel):
(JSC::Wasm::BBQJIT::addBlock):
(JSC::Wasm::BBQJIT::addLoop):
(JSC::Wasm::BBQJIT::addIf):
(JSC::Wasm::BBQJIT::addElse):
(JSC::Wasm::BBQJIT::addElseToUnreachable):
(JSC::Wasm::BBQJIT::addTry):
(JSC::Wasm::BBQJIT::emitCatchPrologue):
(JSC::Wasm::BBQJIT::emitCatchAllImpl):
(JSC::Wasm::BBQJIT::emitCatchImpl):
(JSC::Wasm::BBQJIT::addCatch):
(JSC::Wasm::BBQJIT::addCatchToUnreachable):
(JSC::Wasm::BBQJIT::addCatchAll):
(JSC::Wasm::BBQJIT::addCatchAllToUnreachable):
(JSC::Wasm::BBQJIT::addDelegate):
(JSC::Wasm::BBQJIT::addDelegateToUnreachable):
(JSC::Wasm::BBQJIT::addThrow):
(JSC::Wasm::BBQJIT::addRethrow):
(JSC::Wasm::BBQJIT::prepareForExceptions):
(JSC::Wasm::BBQJIT::addReturn):
(JSC::Wasm::BBQJIT::endBlock):
(JSC::Wasm::BBQJIT::addEndToUnreachable):
(JSC::Wasm::BBQJIT::endTopLevel):
(JSC::Wasm::BBQJIT::flushRegistersForException):
(JSC::Wasm::BBQJIT::returnValuesFromCall):
(JSC::Wasm::BBQJIT::emitCCall):
(JSC::Wasm::BBQJIT::addCall):
(JSC::Wasm::BBQJIT::emitIndirectCall):
(JSC::Wasm::BBQJIT::notifyFunctionUsesSIMD):
(JSC::Wasm::BBQJIT::takeExceptionHandlers):
(JSC::Wasm::BBQJIT::takeCatchEntrypoints):
(JSC::Wasm::BBQJIT::emitShuffle):
(JSC::Wasm::parseAndCompileBBQ):
(JSC::Wasm::BBQJIT::ControlData::allocateArgumentOrResult): Deleted.
* Source/JavaScriptCore/wasm/WasmIRGeneratorHelpers.h:
(JSC::Wasm::computeExceptionHandlerAndLoopEntrypointLocations):

Canonical link: https://commits.webkit.org/260905@main
